src/formats/TNG.cpp

```diff
@@ -71,20 +71,49 @@ TNGFormat::TNGFormat(std::string path, File::Mode mode, File::Compression compre
     CHECK(tng_distance_unit_exponential_get(tng_, &exp));
     // calculate the scale factor from a given length scale to angstrom
     distance_scale_factor_ = pow(10.0, static_cast<double>(exp) + 10.0);
+
+    // work around a bug in tng_num_frames_get (https://redmine.gromacs.org/issues/2937)
+    // manually query all frames in the trajectory, and get the corresponding
+    // TNG frame number in tng_steps_
+    int64_t current_frame = -1;
+    int64_t next_frame = 0;
+    int64_t unused = 0;
+    int64_t* buffer = nullptr;
+
+    tng_function_status status = TNG_SUCCESS;
+    while (true) {
+        // Look for all frames with at least position data
+        int64_t block_ids[] = {TNG_TRAJ_POSITIONS};
+        status = tng_util_trajectory_next_frame_present_data_blocks_find(
+            tng_, current_frame, 1, block_ids, &next_frame, &unused, &buffer
+        );
+
+        if (status == TNG_SUCCESS) {
+            current_frame = next_frame;
+            tng_steps_.push_back(current_frame);
+        } else if (status == TNG_FAILURE) {
+            // We found the end of the file
+            break;
+        } else {
+            // TNG_CRITICAL: throw error
+            CHECK(status);
```

Comment:
I assume this throws then!

Reply:
you are right, I made it more explicit =)